File: app/src/main/java/com/utynote/components/search/SearchContract.java
package com.utynote.components.search;

import android.support.annotation.NonNull;

final class SearchContract {

    public interface View {
        void render(@NonNull SearchViewModel.Abstract model);
    }

    public interface Presenter {

        void attach(@NonNull View view);

        void detach();

        void search(@NonNull String term);
    }
}

File: app/src/main/java/com/utynote/components/search/SearchContract.kt
package com.utynote.components.search

class SearchContract {

    interface View {
        fun render(model: ViewModel)
    }

    interface Presenter {
        fun attach(view: View)

        fun detach()

        fun search(term: String)
    }

    interface ViewModel {

        class Busy : ViewModel

        class Error(val description: String) : ViewModel

        class Data(val data: List<SearchItemData>) : ViewModel
    }
}
